src/admin.c

```diff
@@ -1113,11 +1118,20 @@ static bool admin_cmd_shutdown(PgSocket *admin, const char *arg)
 	 * event from fd.  Currently atexit() cleanup should be called
 	 * before closing open sockets.
 	 */
-	log_info("SHUTDOWN command issued");
-	cf_shutdown = SHUTDOWN_IMMEDIATE;
-	event_base_loopbreak(pgb_event_base);
-
-	return true;
+	cf_shutdown = mode;
+	if (mode == SHUTDOWN_IMMEDIATE) {
+		log_info("SHUTDOWN command issued");
+		event_base_loopbreak(pgb_event_base);
+		/*
+		 * By not running admin_ready the connection is kept open
+		 * until the process is actually shut down.
+		 */
+		return true;
+	} else {
+		log_info("SHUTDOWN WAIT_FOR_CLIENTS command issued");
+		cleanup_tcp_sockets();
```

Comment:
Ah, so in the signal handler you are using `cleanup_sockets()`, but here you use `cleanup_tcp_sockets()`, because you're still connected. See my comments there, we should make this more coherent. I think the behavior we want here is to clean up all connections except *this one*, independent of the socket type.

Reply:
Okay, so I implemented this. One thing I'm wondering is if we should do also close all sockets when issuing `SHUTDOWN WAIT_FOR_SERVERS` (or a SIGINT). It seems quite weird that we're allowing new connections, even though they will be held in waiting state until all the servers are done with their current task. It seems nicer to stop listening for new connections.